# Laboratory excercise 5

The notebook contains excercise connected to the auditory excerice 6. For any questions feel free to contact assistant: eda.jovicic@fer.hr

In [1]:
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf
import pathlib
from sklearn.linear_model import LinearRegression
%matplotlib inline

# set seed for consistency
np.random.seed(2)

1. Load the cleaned dataset from the first exercise. If you haven't saved the dataset, rerun the exercise and save the final dataset.

In [2]:
#load data
org_df = pd.read_csv(f"{str(pathlib.Path().resolve())[:-1]}1\\lab1.csv", index_col=0)
df = org_df.copy()
display(df)

Gender  EthnicGroup  ParentEduc     LunchType  TestPrep  \
0      female            2           1      standard         1   
1      female            2           5      standard         1   
2      female            1           3      standard         1   
3        male            0           0  free/reduced         1   
4        male            2           5      standard         1   
...       ...          ...         ...           ...       ...   
29112  female            3           2      standard         1   
29113    male            4           2      standard         1   
29114  female            2           2  free/reduced         0   
29115  female            3           0      standard         0   
29116    male            1           5      standard         1   

       ParentMaritalStatus  PracticeSport  IsFirstChild  NrSiblings  \
0                        1              1             1         3.0   
1                        1              2             1         0.0   
2                        2              2             1         4.0   
3                        1              0             0         1.0   
4                        1              2             1         0.0   
...                    ...            ...           ...         ...   
29112                    2              2             0         2.0   
29113                    2              1             0         1.0   
29114                    1              2             0         1.0   
29115                    1              1             0         3.0   
29116                    1              0             0         1.0   

       TransportMeans  WklyStudyHours  MathScore  ReadingScore  WritingScore  \
0                   1               1         71            71            74   
1                   2               0         69            90            88   
2                   1               1         87            93            91   
3                   2               0         45            56            42   
4                   1               0         76            78            75   
...               ...             ...        ...           ...           ...   
29112               1               0         59            61            65   
29113               0               0         58            53            51   
29114               0               0         61            70            67   
29115               1               0         82            90            93   
29116               1               0         64            60            58   

       Gender_female  Gender_male  LunchType_free/reduced  LunchType_standard  
0                1.0          0.0                     0.0                 1.0  
1                1.0          0.0                     0.0                 1.0  
2                1.0          0.0                     0.0                 1.0  
3                0.0          1.0                     1.0                 0.0  
4                0.0          1.0                     0.0                 1.0  
...              ...          ...                     ...                 ...  
29112            1.0          0.0                     0.0                 1.0  
29113            0.0          1.0                     0.0                 1.0  
29114            1.0          0.0                     1.0                 0.0  
29115            1.0          0.0                     0.0                 1.0  
29116            0.0          1.0                     0.0                 1.0  

[29117 rows x 18 columns]

2. Using linear regression, model the MathScore using features: TestPrep, NrSiblings, ParentMaritalStatus and ReadingScore. 

In [3]:
# Declares the model

y = df.MathScore
iksevi = ["TestPrep", "NrSiblings", "ParentMaritalStatus", "ReadingScore"]
X = df[iksevi]

X.describe()

TestPrep    NrSiblings  ParentMaritalStatus  ReadingScore
count  29117.000000  29117.000000         29117.000000  29117.000000
mean       0.675276      2.147714             1.109077     69.390081
std        0.468280      1.428635             0.676667     14.760190
min        0.000000      0.000000             0.000000     10.000000
25%        0.000000      1.000000             1.000000     59.000000
50%        1.000000      2.000000             1.000000     70.000000
75%        1.000000      3.000000             2.000000     80.000000
max        1.000000     26.000000             3.000000    100.000000

In [4]:
y.describe()

count    29117.000000
mean        66.559948
std         15.344253
min          0.000000
25%         56.000000
50%         67.000000
75%         78.000000
max        100.000000
Name: MathScore, dtype: float64

In [5]:
X.isna().sum(axis=0)

TestPrep               0
NrSiblings             0
ParentMaritalStatus    0
ReadingScore           0
dtype: int64

In [6]:
model = LinearRegression()
model.fit(X, y)

LinearRegression()

In [7]:
print(str(round(model.intercept_, 3)), end="") # type: ignore
for coeff, col in zip(model.coef_, iksevi):
    print(f" + {str(round(coeff, 3))} \\cdot \\text{{{col}}}", end="")

6.355 + 1.107 \cdot \text{TestPrep} + 0.015 \cdot \text{NrSiblings} + -0.041 \cdot \text{ParentMaritalStatus} + 0.857 \cdot \text{ReadingScore}

$$\text{MathScore} = 6.355 + 1.107 \cdot \text{TestPrep} + 0.015 \cdot \text{NrSiblings} -0.041 \cdot \text{ParentMaritalStatus} + 0.857 \cdot \text{ReadingScore}
$$

In [8]:
jednacina = "MathScore ~ C(TestPrep) + NrSiblings + C(ParentMaritalStatus) + ReadingScore" # naglasene kategoricke varijable operatorom C()
regr = smf.ols(formula=jednacina, data = df).fit() # linearna regresija na danim podacima
regr.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              MathScore   R-squared:                       0.670
Model:                            OLS   Adj. R-squared:                  0.669
Method:                 Least Squares   F-statistic:                     9829.
Date:                Wed, 20 Dec 2023   Prob (F-statistic):               0.00
Time:                        21:17:37   Log-Likelihood:            -1.0471e+05
No. Observations:               29117   AIC:                         2.094e+05
Df Residuals:                   29110   BIC:                         2.095e+05
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
===============================================================================================
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
Intercept                       6.2302      0.314     19.867      0.000       5.616       6.845
C(TestPrep)[T.1]                1.1060      0.113      9.807      0.000       0.885       1.327
C(ParentMaritalStatus)[T.1]     0.1733      0.145      1.193      0.233      -0.112       0.458
C(ParentMaritalStatus)[T.2]    -0.1318      0.168     -0.786      0.432      -0.460       0.197
C(ParentMaritalStatus)[T.3]     0.5432      0.396      1.373      0.170      -0.232       1.319
NrSiblings                      0.0155      0.036      0.428      0.668      -0.055       0.086
ReadingScore                    0.8570      0.004    239.510      0.000       0.850       0.864
==============================================================================
Omnibus:                      233.486   Durbin-Watson:                   1.987
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              155.618
Skew:                          -0.010   Prob(JB):                     1.61e-34
Kurtosis:                       2.642   Cond. No.                         562.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## Questions:

### a) Using the summary, write down the formula with coefficients

$$\mathrm{MathScore} = 6.36 + 1.11 \cdot \mathrm{TestPrep} + 0.0154 \cdot  \mathrm{NrSiblings} -0.041 \cdot  \mathrm{ParentMaritalStatus} + 0.857 \cdot  \mathrm{ReadingScore}$$

### b) Are there any statisticaly important features?

Important features can be detected by looking at the parameters in the summary table-those having the lowest $P>|t|$ values are the most important. If the said parameter is below 0.05, the corresponding feature can be considered important. Those features are **ReadingScore** and **TestPrep**.

### c) Let's interpret the data, if students do test preparation, how does that affect the outcome on the math exam? 

It does help (preparation correlates to higher test scores), but the effect of preparation (1.11 points higher) is not that significant relatively to the total score (100 points).

### d) Are students who practice sport better at math exams or worse?
Using linear regression, model the MathScore using only one feature: PracticeSport. Don't include intercept. Using summary, decide which students do the best at math exams - those who never practice sport, sometimes practice sport or those who do it regularly?

In [9]:
jednacina = "MathScore ~ 0 + C(PracticeSport)" # fiksiramo odsjecak nulom
regr = smf.ols(formula=jednacina, data = df).fit() # linearna regresija na danim podacima
regr.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              MathScore   R-squared:                       0.006
Model:                            OLS   Adj. R-squared:                  0.006
Method:                 Least Squares   F-statistic:                     87.35
Date:                Wed, 20 Dec 2023   Prob (F-statistic):           1.51e-38
Time:                        21:17:37   Log-Likelihood:            -1.2074e+05
No. Observations:               29117   AIC:                         2.415e+05
Df Residuals:                   29114   BIC:                         2.415e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=======================================================================================
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
C(PracticeSport)[0]    64.1576      0.244    262.426      0.000      63.678      64.637
C(PracticeSport)[1]    67.8441      0.150    452.998      0.000      67.551      68.138
C(PracticeSport)[2]    66.2896      0.126    526.520      0.000      66.043      66.536
==============================================================================
Omnibus:                      234.462   Durbin-Watson:                   1.997
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              214.722
Skew:                          -0.171   Prob(JB):                     2.36e-47
Kurtosis:                       2.756   Cond. No.                         1.94
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [10]:
avgs = df.groupby("PracticeSport")["MathScore"].mean()
avgs # za koeficijente u regresiji dobiju se srednje vrijednosti

PracticeSport
0    64.157559
1    67.844082
2    66.289584
Name: MathScore, dtype: float64

Students who do sports are generally better at math than those who don't. However, more sport practices don't lead to better MathScores. **Students who sometimes practice sport** are better than those who do it regularly, thus the answer is the category 1 (sometimes).